### PR TITLE
feat: support using exponential backoff between self heal attempts

### LINKED ISSUE #20172: Self heal exponential delay increase
# Summary

Argo CD makes 5 seconds delay between auto-sync attempts. The delay duration might be configured using `controller.self.heal.timeout.seconds` field of `argocd-cmd-params-cm` ConfigMap. It is proposed to support configuring exponential delay increase.

# Motivation

The delay between attempts is required to prevent unnecessary sync attempts when Out-Of-Sync status is caused by invalid manifests/unexpected K8S controller behavior [etc](https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/). Increasing timeout helps avoid unnecessary controller CPU usage/K8S API server requests but affects end user experience.

The exponential timeout increase would provide both: good user experience and good controller performance. 

# Proposal

Introduce `controller.self.heal.timeout.backoff.factor` `controller.self.heal.timeout.backoff.maxseconds` fields in  `argocd-cmd-params-cm` ConfigMap. If set the new fields and `controller.self.heal.timeout.seconds` should be used to configure exponential timeout increase between auto-sync attempts.

It is proposed to avoid adding new fields and calculate number of attempts based on time since last sync attempt:

```math
n = \log_2 \left( \frac{T_{cumulative}}{T_0} + 1 \right)
```

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/20172

PR implements the ability to use exponential timeout increase between self-heal attempts.  Notes about the implementation:

* PR does introduce a new field in operation state to track number of attempts. I've realized that on each attempt controller updates operation so tracking number of attempts would be free (no additional update API calls )
* PR changes the default controller behavior to exponential backoff (unless custom timeout is set). The reason is exponential timeout increase provides protection from K8S API overload and good user experience at the same time. So no users really need to fine tune it.

### CODE REVIEW

**Review by christianh814 — APPROVED** [review]

**Review by gdsoumya — CHANGES_REQUESTED** [review]

File: controller/appcontroller_test.go [hunk 1]
```diff
@@ -2191,3 +2193,40 @@ func TestAlreadyAttemptSync(t *testing.T) {
...
 	})
 }
+
+func assertDurationAround(t *testing.T, expected time.Duration, actual time.Duration) {
+	delta := time.Second / 2
+	assert.GreaterOrEqual(t, expected, actual-delta)
+	assert.LessOrEqual(t, expected, actual+delta)
+}
+
+func TestSelfHealExponentialBackoff(t *testing.T) {
```
Comment (nitishfy) [inline controller/appcontroller_test.go:2203]:
```suggestion
func TestSelfHealExponentialBackoff(t *testing.T) {
```

This is a great addition. I'm just wondering that we can add some more testcases for this test using the Table Driven approach. WDYT?

File: manifests/base/application-controller-deployment/argocd-application-controller-deployment.yaml [hunk 2]
```diff
@@ -97,6 +97,24 @@ spec:
...
+            configMapKeyRef:
+              name: argocd-cmd-params-cm
+              key: controller.self.heal.backoff.factor
+              optional: true
+        - name: ARGOCD_APPLICATION_CONTROLLER_SELF_HEAL_BACKOFF_CAP_SECONDS
+          valueFrom:
+            configMapKeyRef:
+              name: argocd-cmd-params-cm
+              key: controller.self.heal.backoff.cap.seconds
+              optional: true
```
Comment (nitishfy) [inline manifests/base/application-controller-deployment/argocd-application-controller-deployment.yaml:117]: are these autogenerated?

**Review by nitishfy — COMMENTED** [review]
overall LGTM!

File: controller/appcontroller.go [hunk 3]
```diff
@@ -2120,10 +2127,20 @@ func (ctrl *ApplicationController) shouldSelfHeal(app *appv1.Application) (bool,
...
+		}
 	} else {
-		retryAfter = ctrl.selfHealTimeout - time.Since(app.Status.OperationState.FinishedAt.Time)
+		backOff := *ctrl.selfHealBackOff
+		backOff.Steps = int(app.Status.OperationState.Operation.Sync.SelfHealAttemptsCount)
+		var delay time.Duration
+		for backOff.Steps > 0 {
+			delay = backOff.Step()
+		}
+		retryAfter = delay - time.Since(app.Status.OperationState.FinishedAt.Time)
```
[the remark this answers is not in the document]
Comment (alexmt) [reply controller/appcontroller.go]: ops. good catch! fixed

File: controller/appcontroller_test.go (hunk 1 above)
Comment (alexmt) [reply controller/appcontroller_test.go:2204]: Agree, added more test cases including nil `app.Status.OperationState.FinishedAt`

File: manifests/base/application-controller-deployment/argocd-application-controller-deployment.yaml (hunk 2 above)
Comment (alexmt) [reply manifests/base/application-controller-deployment/argocd-application-controller-deployment.yaml:117]: No, unfortunately we need to manually maintain these mapping

**Review by gdsoumya — APPROVED** [review]

**Pull request merged**